### ISSUE TITLE
Xkcd-style plot flow diagrams

### ISSUE THREAD

**Issue by theq629** [sfu-natlang/lensingwikipedia#111]
(Paraphrased from emails by @anoopsarkar:) As a new view on the frontend, use the entities from the current search to produce a visualization like the ones at http://csclub.uwaterloo.ca/~n2iskand/?page_id=13, which is based on [xkcd #657](http://xkcd.com/657/). The x-axis is like in the current timeline plot but may need more scrolling see detail. The y-axis groups entities (especially persons, but we could also use organizations, etc.) into clusters based on co-occurrence in the same events at any given year, and is sorted by frequency. There is a [d3 plugin](http://bost.ocks.org/mike/sankey/) for Sankey diagrams that should work. We can optionally have the thickness of the flow lines represent frequency, but the initial goal should just be to replicate the Waterloo visualization.

**Comment by anoopsarkar** [comment]
xkcd storyline is live for wikipedia now!

**Comment by KonceptGeek** [comment]
I think it would be a good idea to move the Storyline tab in index.html after Facets.

**Comment by anoopsarkar** [comment]
I agree but let us wait until the merge fest is complete.

**Comment by anoopsarkar** [comment]
There seems to be a corner case bug:

1. select from Timeline 1499 BCE - 1314 BCE (does not have to be exact, just don't select until 1000 BCE).
1. Select "Hatshepsut" from the Person facet.
1. Look at the Storyline tab. It is empty. Selecting "Hatshepsut" without the timeline constraint shows two independent lines. It might be that "Hatshepsut" does not co-occur with any other person.

**Comment by anoopsarkar** [comment]
I'm not sure if this needs a solution but it is one of those strange corner cases. 

If I select a line in Storyline and then remove the original Facet constraint (as I should be able to do in faceted browsing) then there is a constraint from the Storyline still active but the Storyline view insists on a selection from the facet list.  So there is a constraint active but the Storyline view is empty.

**Comment by theq629** [comment]
For the Hatshepsut, another likely possibility is that Hatshepsut occurs in only one storyline cluster in that range. Right now that gets filtered out from the visualization since it can't be drawn as a line. I guess I should add a special case that draws it as a dot on top of the node or something like that.

For the selection case, that does seem confusing. Should the storyline view clear its own constraints at the same time it shows the "Make a selection in the facet" text?

**Comment by anoopsarkar** [comment]
1. a dot sounds good

2. either the constraint is cleared or once a selection is done in the
storyline view then it proceeds independently and shows the storyline view
consistent with the selection.

**Comment by anoopsarkar** [comment]
I moved the Storyline tab in index.html to be right after the Facets tab

**Comment by anoopsarkar** [comment]
should we close this issue and handle bugs or enhancements with separate issues?

**Comment by theq629** [comment]
Why don't I finish the two enhancements above, and then we'll close this issue.

**Comment by anoopsarkar** [comment]
ok. sounds good.

**Comment by anoopsarkar** [comment]
How about simplifying the current back and forth between the Facet view and Storyline view by simply generating a default Storyline for the most frequent element of each facet entry. e.g. currently the Storyline view for the Person facet would be Augustus by default. We can also cache the default view as we do with other tabs.

**Comment by theq629** [comment]
That sounds good, but I'm also wondering if we need to add some sort of introduction message that explains what's being shown, especially if we are introducing arbitrary limits (for the too many entity lines bug, #133). One way to handle that would be to have a help message that is shown initially, like the text search does. But maybe a similar box that doesn't show up until the user asks for it would be find in this case.

**Comment by theq629** [comment]
On the default view, isn't it confusing if there is already a storyline generated before a selection is made?

**Comment by theq629** [comment]
For (2), I've now got it clearing whenever it shows the help text. That also includes when you switch modes (change facets or change to query mode).

For (1), I've added oval dots. Here is the Hatshepsut case:
![storyline-singleton2](https://cloud.githubusercontent.com/assets/1696740/5736045/bfc953b8-9c0b-11e4-9ea4-4210cff87492.png)

There is also a single-cluster node visible if you select Hannibal. I'm not sure this is actually the best way to draw then, but at least we can see all entities now. One concern is that the marker has to extend past the node marker to be selectable

**Comment by anoopsarkar** [comment]
Re: a default view for Storyline. Just like other views have a summary view that is loaded by default, the Storyline can have the most frequent entity in a facet. To make things clearer we can have a drop down menu in the Storyline view that can be used to select the entity (this will also avoid having to switch to the Facet view altogether). 

So the top of Storyline would look like this:

[Clear Selection] [Person facet](*1) [Augustus [439]](*2)

[ Default Storyline view would be Person facet and most frequent Person ]

1: can be used to select other enabled facets (as it is done now)
2: the other elements in this drop down menu would be entities from the selected facet in (1) sorted by frequency. the user can select any of them without switching to the facet view.

One potential issue is that the frequencies and entities will change based on other constraints. But I think that information is readily available already to the Storyline view, isn't it?

**Comment by theq629** [comment]
Ok, I think that would work. Do the selections made in the main facet still change the storyline selection (and what's shown as selected in the storyline menu)?

**Comment by anoopsarkar** [comment]
No, I think we can decouple the Facet view from Storyline view entirely using this approach. That would give us more flexibility in the Facet view (as raised in #138)

**Comment by theq629** [comment]
In that case we'd basically just move a copy of the facet queries to the storyline. So definitely doable, but since it's a pretty big change I suggest we make a new issue for it. (And close this issue if the other two things look ok to you.)

**Comment by anoopsarkar** [comment]
yes, sounds good. please merge your current changes with master.

**Comment by theq629** [comment]
Ok, merged. And I assume that means we can finally close this issue!
